### DISCUSSION TITLE
Fetching JSON multi-level sub-object not supported on `text` column

### DISCUSSION THREAD

**Discussion by yan-hic** [supabase/supabase#12953] [Questions] [answered]
I have json-string data i.e. json-object stored as string, and want to extract sub-object as described at https://www.postgresql.org/docs/15/functions-json.html.

The doc at https://supabase.com/docs/guides/database/json#query-the-jsonb-data does not list a sub-object query but it does not explicitly states it's not supported.

I found 2 scenarios that fail but work fine in SQL:

    sql -> select my-nested-json-field-as-string::json -> 'sub' from table
    js -> ...from('table').select(`my-nested-json-field-as-string::json ->sub`) --> parsing error: `unexpected '-' `
    js -> ...from('table').select(`my-nested-json-field-as-string::json >sub`) --> parsing error: `unexpected '>' `

    sql -> select my-nested-json-field-as-string::json #> '{sub, child}' from table
    js -> ...from('table').select(`my-nested-json-field-as-string::json #>'{sub, child}'`) --> parsing error: `unexpected '#' `

Workaround is to create SQL view in PG but would love if this was supported natively by the lib (any - we use Python SDK too).

**Comment by GaryAustin1** [comment]
In js `.select('my_json_col -> sub')` works and returns the sub json object.
     `.select('my_json_col::json -> sub')`  errors.  You don't normally specify the data type in the REST API.
     
 You mention Python though so maybe that is different, but skeptical.
 You might generate an issue in that repository instead of here, as it maybe specific to that community repo.
 
 It is usually better to seek help in Supabase Discord or Discussions before generating an bug report here.

**Comment by yan-hic** [comment]
I know a json-column works, be it it only works for one-level - I have changed the title to reflect that.
I specifically indicated that the source is a string, not a JSON-type, hence you need to add `::json` to coerce to JSON before applying the qualifier.

The issue exists probably in all libs that use the same parser. The fix - call it enhancement request if you wish - would be to update the regex or alike when parsing to include the 2 scenarios.
Setting the first one aside, the second i.e. querying any-level of a JSON-object is more important in my mind.

Fine with me if a maintainer wants to close it but the limitation does exist and discussing it won't help.

**Comment by GaryAustin1** [comment, marked as answer]
Sorry if I misunderstood what you are trying to do.

For this jsonb column data
```
{
  "sub": {
    "dogs": "three",
    "sub2": {
      "rabbits": "five"
    }
  },
  "cats": "four"
}
```
 `.select('json_col->cats, json_col->sub->>dogs,  json_col->sub->sub2->>rabbits, json_col->sub, json_col->sub->sub2')`
 returns:
![image](https://user-images.githubusercontent.com/54564956/224399158-dffb9c1a-7f92-4982-9108-ca2c851f77ad.png)

Is that not what you mean by accessing sub objects?

And ` #> '{sub, child}'` is NOT supported in the API.

**Reply by yan-hic** [reply]
I didn't realize chaining json operators was possible so you are right on !
```
create table if not exists tbl as select '{"a":{"b":[4,5,6]}}' t, '{"a":{"b":[4,5,6]}}'::json j;`
works: 
...from('tbl').select(`j->a->b->1`);
```
My point was indeed that not all json-operators listed at https://www.postgresql.org/docs/15/functions-json.html were supported but your answer provides a very valid workaround. Thanks !

**Comment by steve-chavez** [comment]
> I specifically indicated that the source is a string, not a JSON-type, hence you need to add ::json to coerce to JSON before applying the qualifier

@yan-hic You don't need to cast to json, `to_json` is used automatically whenever you use the arrow operators. So it should work on a `text` column if it contains a `json`. I would recommend typing the column properly on the db though.

**Reply by yan-hic** [reply]
@steve-chavez `to_json()` has a different use. It's to convert a json-object-as-text to json, not cast. That's what the `::json` (or `cast(xx as json)` is for.
To illustrate:
```
create table if not exists tbl as select '{"a":{"b":[4,5,6]}}' t, '{"a":{"b":[4,5,6]}}'::json j;

fails:
select t->'a' from tbl
select to_json(t)->'a' from tbl

works:
select j->'a' from tbl
select t::json->'a' from tbl
```
I agree passing a json type is ideal but supabase's BigQuery fdw does not support json (yet). 
So one needs to pass string->text and convert the column to json in PG. 

Casting with `::json` would be easier but `postgres_js` (other SDK ?) does not support the notation `from('tbl').select('t::json->a->b->1')` whereas sql does.
However, we have found an easy way around that [here ](https://github.com/supabase/wrappers/issues/70)
